### PR TITLE
Introduce discovery-based client transport

### PR DESCRIPTION
**Pull request by raboof**
References #92

### CODE REVIEW

**Comment by jrudolph** [comment]
Interesting! Why would we use discovery instead of DNS? Is it because we have it and it is more general, so we could e.g. make use of SRV records? Wouldn't it be potentially confusing that the host from the URI is resolved using something else than DNS?

File: akka-http-core/src/main/scala/akka/http/scaladsl/DiscoveryClientTransport.scala [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+                  settings.socketOptions, halfClose = true, settings.connectingTimeout, settings.idleTimeout)
+                  .mapMaterializedValue(_.map(tcpConn => OutgoingConnection(tcpConn.localAddress, tcpConn.remoteAddress))(system.dispatcher))
+            }
+        }
+      }
+    ).mapMaterializedValue(_.flatten)
+
+  }
+
+  private def futureFlow[I, O, M](flow: Future[Flow[I, O, M]]): Flow[I, O, Future[M]] =
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/scaladsl/DiscoveryClientTransport.scala:55]: Could we use the 2.5 `Flow.lazyInitAsync` instead?

Comment (raboof) [reply akka-http-core/src/main/scala/akka/http/scaladsl/DiscoveryClientTransport.scala:55]: I tried that initially. That materializes a future that is resolved when the first element arrives. By flatmapping that future I can keep the `ClientTransport` signature intact, but then a `singleRequest` call stalls. I haven't dug deeper, but perhaps we don't send any elements before the materialized future completes?

**Comment by raboof** [comment]
> Why would we use discovery instead of DNS? Is it because we have it and it is more general, so we could e.g. make use of SRV records?

Yes, that was my thinking.

> Wouldn't it be potentially confusing that the host from the URI is resolved using something else than DNS?

As long as we keep DNS as the default and only use Discovery when explicitly configured I think that might be fine?

**Comment by jrudolph** [comment]
Sounds reasonable.

File: akka-http-core/src/main/scala/akka/http/scaladsl/DiscoveryClientTransport.scala (hunk 1 above)
Comment (raboof) [reply akka-http-core/src/main/scala/akka/http/scaladsl/DiscoveryClientTransport.scala:55]: `NewHostConnectionPool` uses this future to trigger `onConnectionAttemptSucceeded`, after which the slot becomes `Idle` for accepting requests.

Comment (jrudolph) [reply akka-http-core/src/main/scala/akka/http/scaladsl/DiscoveryClientTransport.scala:55]: Makes sense, forgot that `lazyInitAsync` waits for the first element.

**Comment by ennru** [comment]
I added #3202 as a more open variant of this PR.

**Comment by raboof** [comment]
Not sure if this still makes sense now that we have #3202

**Pull request closed without merging**